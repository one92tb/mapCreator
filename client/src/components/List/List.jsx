import React, { Component } from "react";
import { connect } from "react-redux";
import { fetchIndicators } from "../../actions/mapIndicator/fetchIndicators";
import { fetchMarkers } from "../../actions/marker/fetchMarkers";
import { redirectToMain } from "../../actions/redirect/redirect";
import PropTypes from "prop-types";
//import { withRouter } from "react-router-dom";
import {
  Wrapper,
  Label,
  Select,
  Input,
  Form,
  TableContainer,
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  Td,
  Option,
  Image
} from "./style";

Wrapper.displayName = "div";
Label.displayName = "label";
Select.displayName = "select";
Input.displayName = "input";
Form.displayName = "form";
TableContainer.displayName = "div";
Table.displayName = "table";
Thead.displayName = "thead";
Tbody.displayName = "tbody";
Tr.displayName = "tr";
Th.displayName = "th";
Td.displayName = "td";
Option.displayName = "option";
Image.displayName = "img";

export class List extends Component {
  constructor(props) {
    super(props);

    this.state = {
      markerName: "All",
      city: ""
    };
  }

  componentDidMount() {
    const { fetchIndicators, fetchMarkers } = this.props;

    fetchIndicators();
    fetchMarkers();
  }

  handleChange = (e, name) => {
    this.setState({ [name]: e.target.value });
  };

  findIndicatorOnTheMap = indicator => {
    const { redirectToMain } = this.props;
    redirectToMain();
  };

  render() {
    const { markers, indicators, history, redirectToMain } = this.props;
    const { markerName, city } = this.state;

    return (
      <Wrapper>
        <Form>
          <Label>
            <Select onChange={e => this.handleChange(e, "markerName")}>
              <Option>All</Option>
              {markers.map((marker, id, arr) => (
                <Option key={marker.id}>{marker.name}</Option>
              ))};
            </Select>
          </Label>
          <Input
            onChange={e => this.handleChange(e, "city")}
            type="text"
            name="city"
            placeholder="search city"
          />
        </Form>
        <TableContainer>
          <Table>
            <Thead>
              <tr>
                <Th>id</Th>
                <Th>name</Th>
                <Th>street</Th>
                <Th>city</Th>
                <Th>country</Th>
                <Th>find on the map</Th>
              </tr>
            </Thead>
            <Tbody>
              {indicators
                .filter(indicator => {
                  return (markerName === "All" && city === "") ||
                    (markerName === "All" &&
                      indicator.city
                        .toLowerCase()
                        .search(city.toLowerCase()) !== -1)
                    ? indicator
                    : markerName === indicator.name &&
                        indicator.city
                          .toLowerCase()
                          .search(city.toLowerCase()) !== -1 &&
                        indicator;

                  //1 ) sprawdza wszystko ALL ""
                  // 2) All  Wroclaw
                  // 3 ) test1 Wroclaw
                })
                .map((indicator, id) => (
                  <Tr key={indicator.id}>
                    <Td>{id + 1}</Td>
                    <Td>{indicator.name}</Td>
                    <Td>{indicator.street}</Td>
                    <Td>{indicator.city}</Td>
                    <Td>{indicator.country}</Td>
                    <Td>
                      <Image
                        src={"map-location.png"}
                        onClick={() => this.findIndicatorOnTheMap(indicator)}
                      />
                    </Td>
                  </Tr>
                ))}
            </Tbody>
          </Table>
        </TableContainer>
      </Wrapper>
    );
  }
}

const mapDispatchToProps = {
  fetchIndicators,
  fetchMarkers,
  redirectToMain
};

const mapStateToProps = state => ({
  indicators: state.mapIndicator.indicators,
  markers: state.marker.markers
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(List);

List.propTypes = {
  indicators: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
      lat: PropTypes.number.isRequired,
      lng: PropTypes.number.isRequired,
      street: PropTypes.string,
      city: PropTypes.string,
      country: PropTypes.string,
      userId: PropTypes.number.isRequired
    })
  ).isRequired,
  markers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
      userId: PropTypes.number.isRequired
    })
  ),
  fetchIndicators: PropTypes.func.isRequired,
  fetchMarkers: PropTypes.func.isRequired,
  redirectToMain: PropTypes.func.isRequired
};
