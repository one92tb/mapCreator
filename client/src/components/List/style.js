import styled from "styled-components";
import { css } from "styled-components";

const Wrapper = styled.div`
  background: #f2f2f2;
  height: 100%;
  padding: 40px 20px;
`;

const Label = styled.label`
  margin: 0;
`;

const Select = styled.select`
  width: 200px;
  height: 40px;
  color: #000;
  display: block;
  text-align: center;
  border-radius: 5px;
  background: #fff;
  text-align: center;
  text-align-last: center;
`;

const Input = styled.input`
  height: 40px;
  width: 200px;
  text-align: center;
  border-radius: 5px;
  border: 1px solid transparent;
`;

const Form = styled.form`
  height: calc(10% - 20px);
  margin-bottom: 20px;
  display: flex;
  justify-content: space-around;
  background: #00b8e6;
  display: flex;
  align-items: center;
  border-radius: 3px;
`;

const TableContainer = styled.div`
  height: 90%;
`;

const Table = styled.table`
  width: 100%;
  display: block;
  height: 100%;
`;

const Thead = styled.thead`
  display: table;
  width: 100%;
`;

const Tbody = styled.tbody`
  overflow: auto;
  overflow-x: hidden;
  display: block;
  width: 100%;
  height: 100%;
  height: calc(100% - 48px);
  text-align: center;
`;

const Tr = styled.tr`
  display: table;
  width: 100%;
  table-layout: fixed;
  margin: 5px 0;
`;

const Cell = css`
  padding: 0.75rem;
  text-align: center;
  width: 10%;

  &:not(:first-child) {
    width: 15%;
  }
`;

const Th = styled.th`
  ${Cell};
  color: #fff;
  background: #00b8e6;
`;

const Td = styled.td`
  ${Cell};
  background: #4ddbff;
`;

export {
  Wrapper,
  Label,
  Select,
  Input,
  Form,
  TableContainer,
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  Td
};
