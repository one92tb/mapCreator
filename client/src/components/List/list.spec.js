import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { List } from "./List";

describe("list component", () => {
  let wrapper;

  it("passses all props to List component", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: " Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const redirectToMain = jest.fn();
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("tbody with 2 indicators", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );
    expect(wrapper.find("tbody").children().length).toEqual(2);
  });

  it("tbody without indicators", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );
    expect(wrapper.find("tbody").children().length).toEqual(0);
  });

  it("default settings - should return all table body row", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const city = "";
    const markerName = "All";

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    const filteredTr = wrapper
      .find("tbody")
      .children() //[tr,tr]
      .filterWhere(tr => {
        if (markerName === "All" && !city) {
          return tr;
        }
      });
    expect(filteredTr.length).toEqual(2);
  });

  it("filtered table body row by city", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const city = "jelen";
    const markerName = "All";

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    const filteredTr = wrapper
      .find("tbody")
      .children() //[tr,tr]
      .filterWhere(tr => {
        if (
          markerName === "All" &&
          tr
            .childAt(3)
            .children()
            .text()
            .toLowerCase()
            .search(city.toLowerCase()) !== -1
        ) {
          return tr;
        }
      });
    expect(filteredTr.length).toEqual(1);
  });

  it("filtered table body row by markerName and city", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const city = "Wrocław";
    const markerName = "test2";

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    const filteredTr = wrapper
      .find("tbody")
      .children() //[tr,tr]
      .filterWhere(tr => {
        if (
          markerName ===
            tr
              .childAt(1)
              .children()
              .text() &&
          tr
            .childAt(3)
            .children()
            .text()
            .toLowerCase()
            .search(city.toLowerCase()) !== -1
        ) {
          return tr;
        }
      });
    expect(filteredTr.length).toEqual(1);
  });

  it("change marker name in state", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    const select = wrapper.find("select");
    select.simulate("change", { target: { value: "test1" } });
    expect(wrapper.state().markerName).toEqual("test1");
  });

  it("change city in state", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    const input = wrapper.find("input");
    input.simulate("change", { target: { value: "Jelenia Góra" } });

    expect(input.props().name).toEqual("city");
    expect(wrapper.state().city).toEqual("Jelenia Góra");
  });

  it("componentDidMount", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    expect(fetchIndicators).toHaveBeenCalledTimes(1);
    expect(fetchMarkers).toHaveBeenCalledTimes(1);
  });
  it("findIndicatorOnTheMap method have been called 1 time", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];

    const indicator = {
      city: "Jelenia Góra 58-500",
      country: "Polska",
      icon: "1548784752159.png",
      id: 2,
      lat: 50.92841093800815,
      lng: 15.640584390869208,
      name: "test1",
      street: "Unnamed Road",
      userId: 1
    };

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();
    const redirectToMain = jest.fn();

    wrapper = shallow(
      <List
        markers={markers}
        indicators={indicators}
        fetchIndicators={fetchIndicators}
        fetchMarkers={fetchMarkers}
        redirectToMain={redirectToMain}
      />
    );

    //  console.log(wrapper);
    //  console.log(wrapper.instance().props);

    wrapper.instance().findIndicatorOnTheMap = jest.fn(indicator => {
      redirectToMain();
    });

    let { findIndicatorOnTheMap } = wrapper.instance();

    const mapCell = wrapper
      .find("tr")
      .at(1)
      .children()
      .at(5)
      .childAt(0);

    console.log(mapCell.html());
    mapCell.simulate("click");

    expect(findIndicatorOnTheMap).toHaveBeenCalledTimes(1);
    expect(redirectToMain).toHaveBeenCalledTimes(1);
  });
});
