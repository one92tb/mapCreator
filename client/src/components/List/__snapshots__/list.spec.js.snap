// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list component passses all props to List component 1`] = `
<div>
  <form>
    <label>
      <select
        onChange={[Function]}
      >
        <option>
          All
        </option>
        <option
          key="1"
        >
          test1
        </option>
        <option
          key="2"
        >
          basen
        </option>
        ;
      </select>
    </label>
    <input
      name="city"
      onChange={[Function]}
      placeholder="search city"
      type="text"
    />
  </form>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            id
          </th>
          <th>
            name
          </th>
          <th>
            street
          </th>
          <th>
            city
          </th>
          <th>
            country
          </th>
          <th>
            find on the map
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="2"
        >
          <td>
            1
          </td>
          <td>
            test1
          </td>
          <td>
            Unnamed Road
          </td>
          <td>
            Jelenia Góra 58-500
          </td>
          <td>
            Polska
          </td>
          <td>
            <img
              onClick={[Function]}
              src="map-location.png"
            />
          </td>
        </tr>
        <tr
          key="3"
        >
          <td>
            2
          </td>
          <td>
            test1
          </td>
          <td>
            Unnamed Road
          </td>
          <td>
            Jelenia Góra 58-500
          </td>
          <td>
             Polska
          </td>
          <td>
            <img
              onClick={[Function]}
              src="map-location.png"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
